ISPN-13157 file store docs (#9433)
infinispan/infinispan: documentation/src/main/asciidoc/topics/upgrading.adoc

  /v2/caches/{cacheName}/search/indexes?action=reindex

== File-Based Cache Store (`file-store`) Configuration Defaults to SoftIndexFileStore

Including `file-store` persistence in cache configuration now creates a soft index file-based cache store, `SoftIndexFileStore`, instead of a single-file cache store, `SingleFileStore`.
In previous versions `SingleFileStore` was the default for file-based cache stores.

> Any `file-store` configuration is automatically migrated to a `SoftIndexFileStore` after upgrading to Infinispan 13.

Reviewer: ```suggestion
Any `file-store` configuration with a schema version of 13 is automatically migrated to a `SoftIndexFileStore` after upgrading to Infinispan 13.
```

----
ConfigurationBuilder b = new ConfigurationBuilder();
b.persistence()
    .addStore(SoftIndexFileStoreConfigurationBuilder.class)
        .indexLocation("testCache/index");
        .dataLocation("testCache/data")
----

.Infinispan 13 and later
[source,xml,options="nowrap",subs=attributes+,role="secondary"]
----
ConfigurationBuilder b = new ConfigurationBuilder();
b.persistence()
    .addSoftIndexFileStore()
        .indexLocation("testCache/index")
        .dataLocation("testCache/data");
----

=== Migrating Single File Stores

To use `SingleFileStore` as a file-based cache store with Infinispan 13 or later, do one of the following:

.Declarative
[source,xml,options="nowrap",subs=attributes+,role="primary"]
----
<persistence passivation="false">
  <single-file-store shared="false"
                     preload="true"
                     fetch-state="true"
                     read-only="false"/>
</persistence>
----

.Programmatic
[source,java,options="nowrap",subs=attributes+,role="secondary"]
----
ConfigurationBuilder b = new ConfigurationBuilder();
b.persistence()
    .addSingleFileStore()
    .maxEntries(5000);
----

> However, Infinispan does not recommend using `SingleFileStore` because it does not offer the same performance as `SoftIndexFileStore`.

Reviewer: Actually the recommendation is more about scalability since the memory requirements for single file store is much higher than soft index.

b.persistence()
    .addStore(SoftIndexFileStoreConfigurationBuilder.class)
        .indexLocation("testCache/index");
        .dataLocation("testCache/data")
----

.Infinispan 13 and later
[source,xml,options="nowrap",subs=attributes+,role="secondary"]
----
ConfigurationBuilder b = new ConfigurationBuilder();
b.persistence()
    .addSoftIndexFileStore()
        .indexLocation("testCache/index")
        .dataLocation("testCache/data");
----

=== Migrating Single File Stores

To use `SingleFileStore` as a file-based cache store with Infinispan 13 or later, do one of the following:

.Declarative
[source,xml,options="nowrap",subs=attributes+,role="primary"]
----
<persistence passivation="false">
  <single-file-store shared="false"
                     preload="true"
                     fetch-state="true"
                     read-only="false"/>
</persistence>
----

.Programmatic
[source,java,options="nowrap",subs=attributes+,role="secondary"]
----
ConfigurationBuilder b = new ConfigurationBuilder();
b.persistence()
    .addSingleFileStore()
    .maxEntries(5000);
----

However, Infinispan does not recommend using `SingleFileStore`.

> Instead you should use only `SoftIndexFileStore` for file-based cache stores because they offer better performance and scalability.

Reviewer: ```suggestion
Instead you should use `SoftIndexFileStore` for file-based cache stores because it offers better scalability.
```
